admin/javascript/src/components/form-builder/index.js

```diff
@@ -0,0 +1,229 @@
+import GridField from 'components/grid-field/index';
+import fetch from 'isomorphic-fetch';
+
+import es6promise from 'es6-promise';
+es6promise.polyfill();
+
+// Using this to map field types to components until we implement dependency injection.
+const fakeInjector = {
+
+  /**
+   * Components registered with the fake DI container.
+   */
+  components: {
+    TextField,
+    GridField,
+    HiddenField,
+  },
+
+  /**
+   * Gets the component matching the passed component name.
+   * Used when a component type is provided bt the form schema.
+   *
+   * @param string componentName - The name of the component to get from the injector.
+   *
+   * @return object|null
+   */
+  getComponentByName(componentName) {
+    return this.components[componentName];
+  },
+
+  /**
+   * Default data type to component mappings.
+   * Used as a fallback when no component type is provided in the form schema.
+   *
+   * @param string dataType - The data type provided by the form schema.
+   *
+   * @return object|null
+   */
+  getComponentByDataType(dataType) {
+    switch (dataType) {
+      case 'String':
+        return this.components.TextField;
+      case 'Hidden':
+        return this.components.HiddenField;
+      case 'Text':
+        // Textarea field (not implemented)
+        return null;
+      case 'HTML':
+        // HTML editor field (not implemented)
+        return null;
+      case 'Integer':
+        // Numeric field (not implemented)
+        return null;
+      case 'Decimal':
+        // Numeric field (not implemented)
+        return null;
+      case 'MultiSelect':
+        // Radio field (not implemented)
+        return null;
+      case 'SingleSelect':
+        // Dropdown field (not implemented)
+        return null;
+      case 'Date':
+        // DateTime field (not implemented)
+        return null;
+      case 'DateTime':
+        // DateTime field (not implemented)
+        return null;
+      case 'Time':
+        // DateTime field (not implemented)
+        return null;
+      case 'Boolean':
+        // Checkbox field (not implemented)
+        return null;
+      case 'Custom':
+        return this.components.GridField;
+      default:
+        return null;
+    }
+  },
+};
+
+export class FormBuilderComponent extends SilverStripeComponent {
+
+  constructor(props) {
+    super(props);
+
+    this.formSchemaPromise = null;
+    this.isFetching = false;
+
+    this.fetch();
+  }
+
+  /**
+   * Fetches data used to generate a form. This can be form schema and or form state data.
+   * When the response comes back the data is saved to state.
+   *
+   * @param boolean schema - If form schema data should be returned in the response.
+   * @param boolean state - If form state data should be returned in the response.
+   *
+   * @return object - Promise from the AJAX request.
+   */
+  fetch(schema = true, state = false) {
+    const headerValues = [];
+
+    if (this.isFetching === true) {
```

Comment:
Avoiding multiple async calls will likely shift down into `silverstripe-backend`, with the component passing in a promise, but not directly adding promise resolution callbacks. Instead, it'll listen to state updates via `mapStateToProps`. See Sam's asset refactoring: https://github.com/silverstripe/silverstripe-asset-admin/pull/109/files